packages/tupaia-web/src/features/EnlargedDashboardItem/utils/useExportDashboardItem.tsx

```diff
@@ -0,0 +1,185 @@
+  const { type, presentationOptions, name } = config || ({} as DashboardItemConfig);
+  const exportTitle = `${name}, ${entityName}`;
+
+  const { doExport } = useChartDataExport(
+    {
+      ...report,
+      ...config,
+      presentationOptions: {
+        ...(presentationOptions || {}),
+        exportWithLabels,
+        exportWithTable,
+        exportWithTableDisabled,
+      },
+    } as ViewContent,
+    exportTitle,
+  );
+
+  const filename = toFilename(`export-${entityName}-${name}`, true);
+  const file = `${filename}.${exportFormat}`;
+
+  const setExportError = (message: string | null) => {
+    dispatch({ type: ACTION_TYPES.SET_EXPORT_ERROR, payload: message });
+  };
+
+  const setIsExporting = (isExporting: boolean) => {
+    dispatch({ type: ACTION_TYPES.SET_IS_EXPORTING, payload: isExporting });
+  };
+
+  // set the isExporting state to true and the export error state to null when the export starts
+  const handleStartExport = () => {
+    setIsExporting(true);
+    setExportError(null);
+  };
+
+  // set the isExporting state to false and the export error state to null, and download the file when the export is successful
+  const handleExportSuccess = (result: Blob | string) => {
+    downloadJs(result, file);
+    setIsExporting(false);
+    setExportError(null);
+  };
+
+  // set the isExporting state to false and the export error state to the error message when the export fails
+  const handleError = (e: Error) => {
+    setExportError(e.message);
+    setIsExporting(false);
+  };
+
+  // use the useExportToExcel hook to export the matrix dashboard item
+  const { mutate: exportToExcel } = useExportToExcel({
+    onError: (e: Error) => handleError(e),
+    onSuccess: handleExportSuccess,
+    onMutate: handleStartExport,
+  });
+
+  // export to png if the export format is png. This should handle the setting of the isExporting state and the export error state because it is not using a hook from elsewhere
+  const exportToPNG = async () => {
+    try {
+      handleStartExport();
+      const node = exportRef?.current as Node;
+      const result = await domtoimage.toPng(node, { bgcolor: 'white' });
+      handleExportSuccess(result);
+    } catch (e: any) {
+      handleError(e);
+    }
+  };
+
+  const excelParams = getExportToExcelParams({
+    projectCode,
+    dashboardCode: activeDashboard?.code,
+    entityCode,
+    config,
+    report,
+    currentDashboardItem,
+  });
+
+  const EXPORT_FUNCTIONS = {
+    [EXPORT_FORMATS.PNG]: exportToPNG,
+    [EXPORT_FORMATS.XLSX]: type === 'matrix' ? () => exportToExcel(excelParams) : doExport,
+  };
+
+  // set export error message as null and send ga event when the export is successful/starting
+  useEffect(() => {
+    if (exportError || !isExportMode) return;
+    gaEvent('Export', file, isExporting ? 'Attempt' : 'Success');
+  }, [isExporting]);
+
+  // reset the export state when the current dashboard item changes
+  useEffect(() => {
```

Comment:
I am on the fence about this one but I was reading the useEffect docs recently and it recommends a different approach for [resetting state based on props](https://react.dev/learn/you-might-not-need-an-effect#adjusting-some-state-when-a-prop-changes). I'd be interested to hear which you think is best?

Reply:
I'm not sure about that one. I don't really like the idea of having to maintain our own state with previous and current props. It seems like reinventing the wheel, when that's what `useEffect` watches anyway. What do you think?

Reply:
Yes I agree with you that it seems quite clunky to keep the prev props in state so let's stay with the useEffect 👍